### PR TITLE
[BEAM-10322] allow only single assignment to producing stages by pcol…

### PR DESCRIPTION
**Pull request by ihji**
…lection map

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Dataflow | Flink | Samza | Spark
--- | --- | --- | --- | --- | ---
Go | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/)
Python | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/)

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/)
Portable | --- | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by ihji** [comment]
R: @pablom
CC: @robertwb

**Comment by pabloem** [comment]
Run Python2_PVR_Flink PreCommit

**Comment by pabloem** [comment]
can you share how to reproduce the issue, Heejong? @ihji

**Comment by robertwb** [comment]
Could you clarify what the bug that you're trying to fix here is?

File: sdks/python/apache_beam/runners/portability/fn_api_runner/execution.py [hunk 1]
```diff
@@ -375,6 +375,8 @@ def get_all_side_inputs():
         for o in transform.outputs.values():
           if o in s.side_inputs():
             continue
+          if o in producing_stages_by_pcoll:
```
Comment (robertwb) [inline sdks/python/apache_beam/runners/portability/fn_api_runner/execution.py:378]: Why is it OK to ignore duplicates?

**Comment by pabloem** [comment]
We're trying to find producer stages of a PCollection. The first stage
encountered is the producer (due to the topological order) Subsequent
stages encountered with subtransforms producing the PCollection are
consumers where a Data input transform reads in the PCollection.

I can go back to add the comment.

**Comment by ihji** [comment]
Here's a concrete example that can be fixed by this PR:
```
Stage A:
  (Input PCollection i - PTransform 1 - Output PCollection j)
Stage B:
  (Input PCollection j - PTransform 2 - Output PCollection k)
Stage C:
  (Input PCollection j for side input - PTransform 3)
```
We want to find `Stage A` for emitting side input for `Stage C`. However, some synthetic PTransforms are inserted during pipeline optimization phase:
```
Stage A:
  (Input PCollection i - PTransform 1 - Output PCollection j)
  (Input PCollection j - Synthetic Write - Data Sink)
Stage B:
  (Data source - Synthetic Read - Output PCollection j)
  (Input PCollection j - PTransform 2 - Output PCollection k)
Stage C:
  (Input PCollection j for side input - PTransform 3)
```
If we allow multiple assignments to `producing_stages_by_pcoll`, `Stage B` will emit side input for `Stage C` (topologically `Stage B` comes after `Stage A`). Since `Stage B` and `Stage C` have no dependencies, the pipeline will succeed when `Stage B` is executed first and fail when `Stage C` is executed before `Stage B`.

**Comment by aaltay** [comment]
Folks just checking. I do not see an LGTM or an approval on this PR. Is everyone OK with it being merged?

**Comment by pabloem** [comment]
My bad. Heejong and I discussed it via chat last week. I forgot to add the
LGTM when I merged.

**Pull request merged**